docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,142 @@
+check the omap data first as omap data maintained by ceph-csi is not mirrored
+to the remote cluster. This is the challenge ceph-csi needs to handle. If the
+image is not present(means its a local cluster) we need to generate the
+required omap data and store its information and create the rbd image, the CSI
+controller server needs to enable mirroring on the newly created RBD image via
+`rbd mirror image enable image-spec`.
+
+If it is a remote cluster the image will be already created and mirrored by the
+local cluster, we need to check for the image first(if it is present we need to
+generate the omap data and return success to the create volume as its a remote
+cluster we don't need to enable mirroring here).
+
+In case of journal based mirroring the image need to be created with
+exclusive-lock and journaling image features. Additionally, this will also
+imply that `rbd-nbd` mounter will need to be used instead of krbd.
+
+### Promotion/Demotion of the rbd image
+
+During the **ControllerPublish request** from the CSI side, ceph-csi will
+promote the image and use it. and during **ControllerUnpublish request** the
+ceph-csi will demote the image.
+
+as the ControllerPublish is per node so cephcsi need to handle if the image is
+already promoted.
+
+But during **ControllerUnpublish request** the ceph-csi should make sure that
+the rbd image is not attached on any node before demoting the image as it can
+cause data lose. Maybe cephcsi can make use of volume attachment kubernetes
+object to check the pvc is attached to any nodes or not.
+
+If its a ROX or RWX PVC we need to handle image already mirrored case.
+
+#### Incase of dead primary/remote cluster
+
+In case the ceph cluster is dead ceph-csi cannot demote the image to overcome
+this issue we will provide a new flag (configuration) in cephcsi which will be
+set by the high level operator who is responsible for data migration will set
+the flag to `true` so that CSI can forcefully promote the image.
+
+### For snapshot mirroring
+
+This will remain the same as the rbd image mirroring in CreateVolume request,
+here instead of the storageclass parameter, we will check the snapshotclass
+parameter.
+
+same as PVC name and namespace name to generate the rbd image name. we will use
+snapshot name and snapshot namespace name to generate the rbd snapshot name
+and mirror the (intermediate rbd image we create  to the remote cluster),
+created an
+[issue](https://github.com/kubernetes-csi/external-snapshotter/issues/341) in
+external snapshotter to pass extra snapshot metadata to create the snapshot
+request.
+
+### Changes required in cephcsi related to watchers
+
+currently, we are making use of watchers in node stage request we may need to
+change the watchers logic in the node stage request as when we enable the rbd
+mirroring on an image a watcher will be added on an rbd image.
+
+## Work in rook
+
+* Rook may need to provide an option to users (by exposing the CR) to configure
```

Comment:
I think we need clearly identify the items Rook will take care of as well as the one from ceph-csi.

Reply:
I think Rook will be able to provide the Ceph cluster details, I'm just not sure how much configuration it's is going to do. Especially if the replication is at the image level.
They are things like peer additions, I believe ceph-csi should take care of since it knows on which pools it is operating on.